atl-checker/src/algorithms/certain_zero/search_strategy/linear_constrained_phi.rs

```diff
@@ -0,0 +1,253 @@
+                            } else {
+                                LinearConstrainedPhi::Constraint(lin_expr.negated())
+                            }
+                        } else {
+                            LinearConstrainedPhi::NonLinear
+                        };
+                    }
+                    BinaryOpKind::And => {
+                        let lhs_con = self.map_expr_to_constraints(lhs, negated);
+                        let rhs_con = self.map_expr_to_constraints(rhs, negated);
+                        return if !negated {
+                            LinearConstrainedPhi::And(Box::new(lhs_con), Box::new(rhs_con))
+                        } else {
+                            LinearConstrainedPhi::Or(Box::new(lhs_con), Box::new(rhs_con))
+                        };
+                    }
+                    BinaryOpKind::Or => {
+                        let lhs_con = self.map_expr_to_constraints(lhs, negated);
+                        let rhs_con = self.map_expr_to_constraints(rhs, negated);
+                        return if !negated {
+                            LinearConstrainedPhi::Or(Box::new(lhs_con), Box::new(rhs_con))
+                        } else {
+                            LinearConstrainedPhi::And(Box::new(lhs_con), Box::new(rhs_con))
+                        };
+                    }
+                    // P -> Q == not P v Q
+                    BinaryOpKind::Implication => {
+                        let lhs_con = self.map_expr_to_constraints(lhs, !negated);
+                        let rhs_con = self.map_expr_to_constraints(rhs, negated);
+                        return if !negated {
+                            LinearConstrainedPhi::Or(Box::new(lhs_con), Box::new(rhs_con))
+                        } else {
+                            LinearConstrainedPhi::And(Box::new(lhs_con), Box::new(rhs_con))
+                        };
+                    }
+                    _ => {}
+                }
+            }
+            ExprKind::Number(n) => {
+                // Think of != as XOR in this situation
+                return if (*n == 0) != negated {
+                    LinearConstrainedPhi::False
+                } else {
+                    LinearConstrainedPhi::True
+                };
+            }
+            // https://en.wikipedia.org/wiki/Conditioned_disjunction
+            // Q ? P : R == (Q -> P) and (not Q -> R) == (Q and P) or (not Q and R)
+            ExprKind::TernaryIf(q, p, r) => {
+                let q = self.map_expr_to_constraints(q, negated);
+                let p = self.map_expr_to_constraints(p, negated);
+                let r = self.map_expr_to_constraints(r, negated);
+                return if !negated {
+                    LinearConstrainedPhi::Or(
+                        Box::new(LinearConstrainedPhi::And(Box::new(q.clone()), Box::new(p))),
+                        Box::from(LinearConstrainedPhi::And(Box::new(q), Box::new(r))),
```

Comment:
Should'nt q be negated in the second `And` inside the `Or`?

Reply:
![image](https://user-images.githubusercontent.com/33687173/118358195-08053d80-b57e-11eb-9d5f-bbe67690b011.png)

The top table is what the ternary truth table is, and bottom is what I understand it is being translated to. Am I understanding the code wrongly?

Reply:
Good catch!

Reply:
Visuals made by paint (not .net) gang